lib/AST/ASTMangler.cpp

```diff
@@ -2346,30 +2346,83 @@ void ASTMangler::appendContext(const DeclContext *ctx, StringRef useModuleName)
+    // 2. The extension is to a protocol.
+    // 3. The extension has a different generic signature than the extended
+    //    type's generic signature OR if they were the same, if there are any
+    //    inverse requirements imposed.
+    //
     // FIXME: In a world where protocol extensions are dynamically dispatched,
     // "extension is to a protocol" would no longer be a reason to use the
     // extension mangling, because an extension method implementation could be
     // resiliently moved into the original protocol itself.
-      auto sig = ExtD->getGenericSignature();
-      // If the extension is constrained, mangle the generic signature that
-      // constrains it.
-      appendAnyGenericType(decl);
-      appendModule(ExtD->getParentModule(), useModuleName);
-      if (sig && ExtD->isConstrainedExtension()) {
-        Mod = ExtD->getModuleContext();
-        auto nominalSig = ExtD->getSelfNominalTypeDecl()
-                            ->getGenericSignatureOfContext();
-        appendGenericSignature(sig, nominalSig);
+
+    auto sig = ExtD->getGenericSignature();
+    auto nominalSig = ExtD->getSelfNominalTypeDecl()
+                        ->getGenericSignatureOfContext();
+
+    auto mangleExtension = [&](bool genericSignatureRequiredMangling) {
+      if (!ExtD->isInSameDefiningModule() ||
+          ExtD->getDeclaredInterfaceType()->isExistentialType() ||
+          genericSignatureRequiredMangling) {
+        appendAnyGenericType(decl);
+        appendModule(ExtD->getParentModule(), useModuleName);
+
+        if (sig && genericSignatureRequiredMangling) {
+          appendGenericSignature(sig, nominalSig);
+        }
+
+        return appendOperator("E");
+      }
+
+      return appendAnyGenericType(decl);
+    };
+
+    // If we don't have a generic signature, then we don't need to go through
+    // the work of determining generic signature differences or inverse
+    // requirements and such.
+    if (!sig) {
+      return mangleExtension(/* genericSignatureRequiredMangling */ false);
+    }
+
+    SmallVector<Requirement, 2> reqs;
+    SmallVector<InverseRequirement, 2> inverseReqs;
+    sig->getRequirementsWithInverses(reqs, inverseReqs);
+
+    // If the extension's generic signature == the extended type's generic
+    // signature, then we only we need to check if there are any inverse
+    // requirements present.
+    if (sig->isEqual(decl->getGenericSignature())) {
+      // Mangle the extension if there are inverse requirements present.
+      return mangleExtension(
+          /* genericSignatureRequiredMangling */ !inverseReqs.empty());
+    }
+
+    // Otherwise, we need to check if all of the requirements not satisfied by
+    // the context decl are copyable requirements. If all of them are copyable
+    // requirements, then we don't need to mangle the generic signature 
+    auto requirementDiff = sig.requirementsNotSatisfiedBy(nominalSig);
+
+    for (auto req : requirementDiff) {
```

Comment:
This might not be quite right if you have a mix of generic parameters that are Copyable or not, on the extended type and/or extension?